src/rules/cognitive-complexity.ts

```diff
@@ -0,0 +1,306 @@
+      nestingNodes.add(conditionalExpression.consequent);
+      nestingNodes.add(conditionalExpression.alternate);
+    }
+
+    function visitLogicalExpression(logicalExpression: estree.LogicalExpression) {
+      if (!logicalOperationsToIgnore.has(logicalExpression)) {
+        const flattenedLogicalExpressions = flattenLogicalExpression(logicalExpression);
+
+        let previous: estree.LogicalExpression | undefined;
+        for (const current of flattenedLogicalExpressions) {
+          if (!previous || previous.operator !== current.operator) {
+            addComplexity();
+          }
+          previous = current;
+        }
+      }
+    }
+
+    function flattenLogicalExpression(node: estree.Node): estree.LogicalExpression[] {
+      if (isLogicalExpression(node)) {
+        logicalOperationsToIgnore.add(node);
+        return [...flattenLogicalExpression(node.left), node, ...flattenLogicalExpression(node.right)];
+      }
+      return [];
+    }
+
+    function addStructuralComplexity() {
+      const added = nesting + 1;
+      if (enclosingFunctions.length === 1) {
+        // top level function
+        topLevelHasStructuralComplexity = true;
+        topLevelOwnComplexity += added;
+      } else {
+        complexityIfNested += added + 1;
+        complexityIfNotNested += added;
+      }
+    }
+
+    function addComplexity() {
+      if (enclosingFunctions.length === 1) {
+        // top level function
+        topLevelOwnComplexity++;
+      } else {
+        // second+ level function
+        complexityIfNested++;
+        complexityIfNotNested++;
+      }
+    }
+
+    function checkFunction(node: FunctionLike, complexity: number, loc: OptionalLocation) {
+      if (complexity > threshold) {
+        const location = loc ? { loc } : { node };
+        context.report({
+          message: MESSAGE,
+          data: { complexity: String(complexity), threshold: String(threshold) },
+          ...location,
+        });
+      }
+    }
+
+    /**
+     * Returns
+     * - function name token for function declarations, methods and accessors
+     * - "function" keyword for function expressions
+     * - "=>" for arrow functions
+     */
+    function getLocationToReport(functionLike: FunctionLike): OptionalLocation {
+      switch (functionLike.type) {
+        case "FunctionDeclaration":
+          // `FunctionDeclaration.id` can be null when it is `export default function` (despite of the @types/estree definition)
+          if (functionLike.id) {
+            return functionLike.id.loc;
+          } else {
+            const token = context.getSourceCode().getFirstToken(functionLike);
+            return token && token.loc;
+          }
+        case "FunctionExpression":
+          const parent = getParent(context);
+          if (parent && parent.type === "MethodDefinition") {
+            return parent.key.loc;
+          } else {
+            const token = context.getSourceCode().getFirstToken(functionLike);
```

Comment:
@stas-vilchik I don't like these `getFirstToken` here, it can be `async`
 ```javascript
var functionExpression = async function(a, b) { return a && b; }
```

How about 
```typescript
const token = context.getSourceCode().getTokens(functionLike).find(token => token.value === "function");
```